client/network/src/protocol/notifications/service/mod.rs

```diff
@@ -0,0 +1,468 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Notification service implementation.
+
+use crate::{
+	error,
+	protocol::notifications::handler::NotificationsSink,
+	service::traits::{NotificationEvent, NotificationService, ValidationResult},
+	types::ProtocolName,
+};
+
+use futures::{
+	stream::{FuturesUnordered, Stream},
+	StreamExt,
+};
+use libp2p::PeerId;
+use tokio::sync::{mpsc, oneshot};
+use tokio_stream::wrappers::ReceiverStream;
+
+use sc_network_common::role::ObservedRole;
+use sc_utils::mpsc::{tracing_unbounded, TracingUnboundedReceiver, TracingUnboundedSender};
+
+use std::{
+	collections::HashMap,
+	fmt::Debug,
+	sync::{Arc, Mutex},
+};
+
+mod metrics;
+#[cfg(test)]
+mod tests;
+
+/// Logging target for the file.
+const LOG_TARGET: &str = "notification-service";
+
+/// Type representing subscribers of a notification protocol.
+type Subscribers = Arc<Mutex<Vec<TracingUnboundedSender<InnerNotificationEvent>>>>;
```

Comment:
The size can change during the "initialization phase", i.e., when GRANDPA tests are creating the voter and observer for its tests. Theoretically it can change during runtime because you can call `NotificationService::clone()` at any point but in practice it's not used anywhere during runtime.